### DISCUSSION TITLE
Drop dependency on npm for C code generation?

### DISCUSSION THREAD

**Discussion by mjambon** [tree-sitter/tree-sitter#1413] [Ideas] [answered]
tl;dr grammars don't need to depend on npm packages, let's eliminate these dependencies to simplify and speed things up.

The `tree-sitter` command, a rust executable, generates C parsers from either a `grammar.js` file and its dependencies or from a `grammar.json` file, and optionally from a `scanner.c` or `scanner.cc` file. JavaScript is required to translate a source grammar `grammar.js` into its final representation `grammar.json`.

The use of Node.js and the dependency on a large number of npm packages to define grammars has been problematic for us, and so repeatedly, since we started using it one and a half year ago. In the ocaml-tree-sitter project, which we use for semgrep, we have currently [21 tree-sitter parsers](https://github.com/returntocorp/ocaml-tree-sitter-semgrep/tree/main/lang) and the list is growing. We extend each grammar with semgrep-specific constructs and process the json grammar further to obtain a fully typed OCaml interface to the CST. This is a rather complex pipeline.

The persistent problems we have are the following:
* installing npm packages is slow
* the build is complicated
* the build breaks often
* as a result, we have to [pre-generate code and pretend it's source code](https://github.com/returntocorp/ocaml-tree-sitter-semgrep/blob/main/doc/release.md#appendix-list-of-semgrep--repos)
* me and my colleagues who work on program analysis unfortunately have no expertise in Node.js and no interest in using it.

In general, it's unclear if all the parsers can be built in the same node environment since each defines its own npm dependencies and version constraints.

Coming from a world of yacc-like parsers (in our case, menhir), we expect code generation for a grammar to take less than a second and to be part of our regular build system. This is not realistic at the moment due to slowness (~20 min to generate everything with our pipeline for all the languages).

So, this post is primarily to say that I don't enjoy the dependency on the Node ecosystem for parser generation. I see it as superfluous since grammar definitions shouldn't have to depend on external libraries.

If I were to do something about it, the options that come to mind are:
1. keep JavaScript for the grammar DSL but disallow dependency on external packages (anything outside of the local repo).
2. implement a new DSL from scratch.

What does everyone think?

**Comment by ahelwer** [comment]
Have you done any profiling to determine where the bulk of time is spent when generating parser code? My understanding is that LR(1) parsers have a comparatively large size compared to other types of parsers, so it might just be that.

**Reply by mjambon** [reply]
It's not just due to the sheer size of the output (parser.c) because I see big differences from one grammar to another:
* typescript: [parser.c is 8.5 MB](https://github.com/returntocorp/semgrep-typescript/blob/36ad9c2459434722bd023fc930a6c77676bf8e02/lib/parser.c), takes 9 s to generate from grammar.json
* hack: [parser.c is 10.1 MB](https://github.com/returntocorp/semgrep-hack/blob/b130715ba3fdc6ae4c2af7cf8d213a62d5c6f08c/lib/parser.c), takes 4 min to generate from grammar.json and uses 15 GB of memory.

I should make a proper report and/or try to figure out where CPU time (and memory) goes.

Note these `grammar.json` files are the result of transformations, which include exposing all the hidden rules (~ removing the leading underscores). Processing the original grammar (`grammar.js`) is typically faster, but there are also discrepancies from one language to another.

**Comment by maxbrunsfeld** [comment, marked as answer]
I don't understand the problem; NPM is not used in any way when generating parsers. There's no need to even have it installed.

You *do* need to have a `node` executable on your PATH (but the *version* of Node doesn't matter at all). Node is just used for converting JavaScript grammar syntax into JSON, but that process is instantaneous; it has nothing to do with the time it takes to generate some parsers.

If you want, you *can* use `npm` as a convenience for installing dependent grammars, in the relatively rare case where one grammar depends on another grammar (like TypeScript depends on JavaScript), but you don't even need NPM for that. All you need to do is set your `NODE_PATH` environment variable to some folder that contains the upstream repo (e.g. `tree-sitter-javascript`). NPM is a simple script that downloads packages into a `./node_modules` folder, which is part of the default `NODE_PATH`.

Maybe you could clarify further what actual problem Node is causing? I'm not seeing how it adds much friction.

**Reply by maxbrunsfeld** [reply]
Some parsers do take a while to generate. The longest one that I know of is `tree-sitter-c-sharp`, which is around 16 seconds on my laptop. Most of the other grammars that I've worked on are in the 0-6 second range though. This time is all spent in the parser-generation algorithm itself though. I've put a fair amount of work into optimizing that, but it can always be improved. Like I said though, it has nothing to do with JavaScript or NPM.

**Reply by mjambon** [reply]
Thanks for the reply. I know my question was vague and lacks details to do much about it... It's just that I was encountering problems while trying to integrate tree-sitter-elixir work with ocaml-tree-sitter. Something was broken and it was the end of the week and I was frustrated.

So, it would be awesome if I managed to set things up for us such that basically everything is "fast and robust". Right now we do install npm packages and so far I wasn't understanding what is really necessary and what isn't. I'll try a few things based on what you suggested and will report more details.

Regarding the parser generator itself, we experience some extra slowness and high memory consumption specifically after converting all the hidden rules to non-hidden rules i.e. removing the leading underscore (e.g. `_expression` &rarr; `expression`). I have to file a proper issue about that.

**Comment by theHamsta** [comment]
@mjambon Is there any reason why you must generate the parsers? At github.com/nvim-treesitter/nvim-treesitter we compile the parsers from the C files which most parsers have in their repo. This way we don't even have to have a node dependency for most parsers.

**Reply by mjambon** [reply]
There are several reasons why we need to generate parser code ourselves:
* We extend the grammars to support [semgrep constructs](https://semgrep.dev/docs/writing-rules/pattern-syntax/).
* We need every rule name to be exposed in the CST as a prerequisite to [recovering a typed tree](https://github.com/returntocorp/ocaml-tree-sitter-core/blob/main/doc/parsing.md). So, every hidden rule like `_foo` must be renamed into `foo`. This is done by rewriting the `grammar.json` file that's produced by `tree-sitter generate` in a first pass. We then feed `grammar.json` to `tree-sitter generate` to produce the final C parser.
* Generated files such as `src/parser.c` are large and in general escape ~~human~~ review. This could allow an external contributor to sneak in some malicious code. The solution is to generate code ourselves.

**Comment by maxbrunsfeld** [comment]
Also related: https://github.com/tree-sitter/tree-sitter/issues/465#issuecomment-773638049. We've talked about adding either:
* An embedded JavaScript interpreter in the CLI, so  you don't need to have a `node` executable
* An environment variable that makes Tree-sitter use a *different* JavaScript interpreter, to avoid requiring that it be Node.js

**Comment by oovm** [comment]
I am binding the backend of my grammar description DSL to tree-sitter.

I am not using node to preprocessing `grammar.js` -> `grammar.json`

My method is based on this step, and I generate `InputGrammar` by myself

https://github.com/tree-sitter/tree-sitter/blob/fb6bd2cad3778bc8cc4c1eef053d56fba4a5040e/cli/src/generate/mod.rs#L68

First, I split `generate_parser_in_directory` into two parts, and then the following definition was leaked to the public:

```rs
use tree_sitter_cli::{
    generate::grammars::{InputGrammar, Variable, PrecedenceEntry, VariableType},
    generate::rules::{Rule, MetadataParams, Associativity, Alias, Symbol, SymbolType},
}
```

---

Or there is a more suitable way for cli is to allow incoming custom DSL path here:

https://github.com/tree-sitter/tree-sitter/blob/fb6bd2cad3778bc8cc4c1eef053d56fba4a5040e/cli/src/generate/mod.rs#L61

Then allow to register a hook here to customize the generator:

https://github.com/tree-sitter/tree-sitter/blob/fb6bd2cad3778bc8cc4c1eef053d56fba4a5040e/cli/src/generate/mod.rs#L160-L169

**Comment by mjambon** [comment]
Thank you all for your answers and suggestions. I finally got us to transition to a setup without npm, which uses just Node.js and setting the `NODE_PATH` environment variable appropriately to manage local dependencies.

**Comment by alemuller** [comment]
Joining the discussion a bit late.

**About the performance.**

A while ago someone was trying to generate an Ada parser with a grammar used in another parser (search for Ada on issues), but tree-sitter got stuck in the generation. I had the same issue while implementing the VHDL parser (that is derived from Ada).

In my case, it happened when trying to implement the expression as it is define in the language reference manual:

https://user-images.githubusercontent.com/12538856/139020992-7cd594f1-bd9d-4d55-8c0d-95278c53660c.png

A grammar containing only this expression does generate a parse quite fast. But when it is inside a larger grammar, slows down the generator significantly.

Based on my experience. The expression (including the primaries) seems to be the most contributing factor for the size and time to generate the parser. As the OP already seems to have realized.

**About using other lexers**

It would be nice if Tree-sitter adopted the r2c API -- see [r2c C manual](https://re2c.org/manual/manual_c.html). This would vastly simplify writing complex external scanners.

**Comment by shi-yan** [comment]
perhaps, just use https://github.com/theduke/quickjs-rs to replace nodejs
